pkg/eestream/encode.go

```diff
@@ -38,71 +38,103 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	r       io.Reader
+	es      ErasureScheme
+	inbuf   []byte
+	outbufs [][]byte
+	chans   map[int]chan []byte
+	err     error
 }
 
 // EncodeReader will take a Reader and an ErasureScheme and return a slice of
 // Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+func EncodeReader(r io.Reader, es ErasureScheme, maxBufferMemory int) []io.Reader {
 	er := &encodedReader{
 		r:       r,
 		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
 		inbuf:   make([]byte, es.DecodedBlockSize()),
 		outbufs: make([][]byte, es.TotalCount()),
+		chans:   make(map[int]chan []byte, es.TotalCount()),
 	}
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
 		readers = append(readers, &encodedPiece{
 			er: er,
 			i:  i,
 		})
 	}
+	if maxBufferMemory < 0 {
+		er.err = Error.New("negative max buffer memory")
+		return readers
+	}
+	chanSize := maxBufferMemory / (es.TotalCount() * es.EncodedBlockSize())
```

Comment:
nit - could use https://golang.org/pkg/math/#Ceil

Reply:
It won't catch the case of maxBufferMemory = 0. Ceil will return 0 and I will still need to additional check to increment it to 1.